pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/LookupTransformFunction.java

```diff
@@ -0,0 +1,332 @@
+
+    _lookupColumnFieldSpec = _dataManager.getColumnFieldSpec(_dimColumnName);
+    Preconditions.checkArgument(_lookupColumnFieldSpec != null,
+        String.format("Column does not exist in dimension table: %s:%s", _dimTableName, _dimColumnName));
+
+    for (String joinKey : _joinKeys) {
+      FieldSpec pkColumnSpec = _dataManager.getColumnFieldSpec(joinKey);
+      Preconditions.checkArgument(pkColumnSpec != null,
+          String.format("Primary key column doesn't exist in dimension table: %s:%s", _dimTableName, joinKey));
+      _joinValueFieldSpecs.add(pkColumnSpec);
+    }
+
+    List<String> tablePrimaryKeyColumns = _dataManager.getPrimaryKeyColumns();
+    Preconditions.checkArgument(_joinKeys.equals(tablePrimaryKeyColumns),
+        String.format("Provided join keys (%s) must be the same as table primary keys: %s", _joinKeys,
+            tablePrimaryKeyColumns));
+  }
+
+  @Override
+  public TransformResultMetadata getResultMetadata() {
+    return new TransformResultMetadata(_lookupColumnFieldSpec.getDataType(),
+        _lookupColumnFieldSpec.isSingleValueField(), false);
+  }
+
+  private Object[] lookup(ProjectionBlock projectionBlock) {
+    int numPkColumns = _joinKeys.size();
+    int numDocuments = projectionBlock.getNumDocs();
+    Object[][] pkColumns = new Object[numPkColumns][];
+    for (int i = 0; i < numPkColumns; i++) {
+      FieldSpec.DataType colType = _joinValueFieldSpecs.get(i).getDataType();
+      TransformFunction tf = _joinValueFunctions.get(i);
+      switch (colType) {
+        case STRING:
+          pkColumns[i] = tf.transformToStringValuesSV(projectionBlock);
+          break;
+        case INT:
+          pkColumns[i] = ArrayUtils.toObject(tf.transformToIntValuesSV(projectionBlock));
+          break;
+        case LONG:
+          pkColumns[i] = ArrayUtils.toObject(tf.transformToLongValuesSV(projectionBlock));
+          break;
+        case FLOAT:
+          pkColumns[i] = ArrayUtils.toObject(tf.transformToFloatValuesSV(projectionBlock));
+          break;
+        case DOUBLE:
+          pkColumns[i] = ArrayUtils.toObject(tf.transformToDoubleValuesSV(projectionBlock));
+          break;
+        case BYTES:
+          byte[][] primitiveValues = tf.transformToBytesValuesSV(projectionBlock);
+          pkColumns[i] = new Byte[numDocuments][];
+          for (int n = 0; n < numDocuments; n++) {
+            pkColumns[i][n] = ArrayUtils.toObject(primitiveValues[n]);
+          }
+          break;
+        default:
+          throw new IllegalStateException("Unknown column type for primary key");
+      }
+    }
+
+    Object[] resultSet = new Object[numDocuments];
+    for (int i = 0; i < numDocuments; i++) {
+      // prepare pk
+      Object[] pkValues = new Object[numPkColumns];
+      for (int j = 0; j < numPkColumns; j++) {
+        pkValues[j] = pkColumns[j][i];
+      }
+      // lookup
+      GenericRow row = _dataManager.lookupRowByPrimaryKey(new PrimaryKey(pkValues));
```

Comment:
`PrimaryKey` object can also be reused

Reply:
Replied [above](https://github.com/apache/incubator-pinot/pull/6383#discussion_r549837013)